pkg/ccl/changefeedccl/sink_pubsub.go

```diff
@@ -0,0 +1,526 @@
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"context"
+	"encoding/json"
+	"hash/crc32"
+	"net/url"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+const credentialsParam = "CREDENTIALS"
+const gcpScheme = "gcppubsub"
+const gcpScope = "https://www.googleapis.com/auth/pubsub"
+
+// TODO: make numOfWorkers configurable
+const numOfWorkers = 128
+
+// isPubsubSInk returns true if url contains scheme with valid pubsub sink
+func isPubsubSink(u *url.URL) bool {
+	switch u.Scheme {
+	case gcpScheme:
+		return true
+	default:
+		return false
+	}
+}
+
+//  i dont really like having getTopicName in the interface but fullname test requires it
+type pubsubClient interface {
+	openTopics(string) error
+	closeTopics(string)
+	flushTopics()
+	sendMessage([]byte, descpb.ID, string) error
+	sendMessageToAllTopics([]byte, string) error
+	getTopicName(descpb.ID) string
+}
+
+// payload struct is sent to the sink
+type payload struct {
+	Key   json.RawMessage `json:"key"`
+	Value json.RawMessage `json:"value"`
+	Topic string          `json:"topic"`
+}
+
+// pubsubMessage is sent to worker channels for workers to consume
+type pubsubMessage struct {
+	alloc   kvevent.Alloc
+	message payload
+	isFlush bool
+	topicID descpb.ID
+}
+
+type gcpPubsubClient struct {
+	client    *pubsub.Client
+	topics    map[descpb.ID]*topicStruct
+	ctx       context.Context
+	projectID string
+	region    string
+	url       sinkURL
+}
+
+type topicStruct struct {
+	topicName   string
+	topicClient *pubsub.Topic
+}
+
+type pubsubSink struct {
+	numWorkers int
+
+	workerCtx   context.Context
+	workerGroup ctxgroup.Group
+
+	exitWorkers func()               // Signaled to shut down all workers.
+	eventsChans []chan pubsubMessage //channel where messages are consumed and sent out
+
+	// flushDone channel signaled when flushing completes.
+	flushDone chan struct{}
+
+	// errChan is written to indicate an error while sending message.
+	errChan chan error
+
+	withTopicName string
+
+	client pubsubClient
+}
+
+// my idea is to force this to be an interface so it adheres to api, but not too sure since Sink is already an API
+//func getPubsubSinkObject(url sinkURL, withTopicName bool) pubsubSink {
+//	return &pubsubSinkObject{url: url, withTopicName: withTopicName}
+//}
+
+// getGCPCredentials returns gcp credentials parsed out from url
```

Comment:
nit: Let's add a TODO to unify this code with basically whatever cloudstorage is doing wrt to GCP credentials parsing.